observer/baseNodeProvider.go

```diff
@@ -309,30 +319,56 @@ func (bnp *baseNodeProvider) removeNodeUnprotected(node *data.NodeData) {
 }
 
 func (bnp *baseNodeProvider) removeNodeFromSyncedNodesUnprotected(nodeToRemove *data.NodeData) {
-	source := &bnp.syncedNodes
 	if nodeToRemove.IsFallback {
-		source = &bnp.syncedFallbackNodes
+		bnp.removeFallbackFromSyncedListUnprotected(nodeToRemove)
+		return
 	}
 
-	bnp.removeNodeFromListUnprotected(nodeToRemove, source)
+	bnp.removeRegularFromSyncedListUnprotected(nodeToRemove)
 }
 
-func (bnp *baseNodeProvider) removeNodeFromListUnprotected(nodeToRemove *data.NodeData, source *[]*data.NodeData) {
-	nodeIndex := -1
-	for idx, node := range *source {
-		if node.Address == nodeToRemove.Address && node.ShardId == nodeToRemove.ShardId {
-			nodeIndex = idx
-			break
-		}
+func (bnp *baseNodeProvider) removeRegularFromSyncedListUnprotected(nodeToRemove *data.NodeData) {
+	nodeIndex := getIndexFromList(nodeToRemove, bnp.syncedNodes)
+	if nodeIndex == -1 {
+		return
 	}
 
+	copy(bnp.syncedNodes[nodeIndex:], bnp.syncedNodes[nodeIndex+1:])
+	bnp.syncedNodes[len(bnp.syncedNodes)-1] = nil
+	bnp.syncedNodes = bnp.syncedNodes[:len(bnp.syncedNodes)-1]
+}
+
+func (bnp *baseNodeProvider) removeFallbackFromSyncedListUnprotected(nodeToRemove *data.NodeData) {
+	nodeIndex := getIndexFromList(nodeToRemove, bnp.syncedFallbackNodes)
 	if nodeIndex == -1 {
 		return
 	}
 
-	copy((*source)[nodeIndex:], (*source)[nodeIndex+1:])
-	(*source)[len(*source)-1] = nil
-	*source = (*source)[:len(*source)-1]
+	copy(bnp.syncedFallbackNodes[nodeIndex:], bnp.syncedFallbackNodes[nodeIndex+1:])
+	bnp.syncedFallbackNodes[len(bnp.syncedFallbackNodes)-1] = nil
+	bnp.syncedFallbackNodes = bnp.syncedFallbackNodes[:len(bnp.syncedFallbackNodes)-1]
+}
+
+func (bnp *baseNodeProvider) removeRegularFromOutOfSyncListUnprotected(nodeToRemove *data.NodeData) {
+	nodeIndex := getIndexFromList(nodeToRemove, bnp.outOfSyncNodes)
+	if nodeIndex == -1 {
+		return
+	}
+
+	copy(bnp.outOfSyncNodes[nodeIndex:], bnp.outOfSyncNodes[nodeIndex+1:])
+	bnp.outOfSyncNodes[len(bnp.outOfSyncNodes)-1] = nil
+	bnp.outOfSyncNodes = bnp.outOfSyncNodes[:len(bnp.outOfSyncNodes)-1]
+}
+
+func (bnp *baseNodeProvider) removeFallbackFromOutOfSyncListUnprotected(nodeToRemove *data.NodeData) {
+	nodeIndex := getIndexFromList(nodeToRemove, bnp.outOfSyncFallbackNodes)
+	if nodeIndex == -1 {
+		return
+	}
+
+	copy(bnp.outOfSyncFallbackNodes[nodeIndex:], bnp.outOfSyncFallbackNodes[nodeIndex+1:])
+	bnp.outOfSyncFallbackNodes[len(bnp.outOfSyncFallbackNodes)-1] = nil
+	bnp.outOfSyncFallbackNodes = bnp.outOfSyncFallbackNodes[:len(bnp.outOfSyncFallbackNodes)-1]
 }
```

Comment:
here instead of these 4 very similar functions we could have only one with 2 params: nodeToRemove and nodes list
but can be also done in a future refactoring if we plan that